jplag/src/test/java/de/jplag/special/TokenPrinterTest.java

```diff
@@ -64,6 +65,21 @@ void printKotlinFiles() {
         printSubmissions(options -> options.setLanguageOption(LanguageOption.KOTLIN));
     }
 
+    @Test
+    void printFilesAsSubmissions() {
+        try {
+            JPlagResult result = runJPlag(FILES_AS_SUBMISSIONS, options -> {
+            });
+            for (Submission submission : result.getSubmissions().getSubmissions()) {
+                printSubmission(submission);
+            }
+            System.out.println("JPlag printed " + result.getSubmissions().numberOfSubmissions() + " valid submissions!");
+        } catch (ExitException exception) {
+            System.err.println("JPlag threw Error: " + exception.getMessage());
+            fail();
+        }
+    }
+
```

Comment:
This is not ideal, as this class is not a test class in the traditional sense. If we want to unit test the token printer, we should actually do that in the module where the token printer resides. Also, no need to run JPlag to print tokens, you can just use the frontend itself and pass the tokens directly to the printer. This makes the unit test more fine-grained.

Reply:
So, should TokenPrinterTest.java just be moved to the package de.jplag, or should a separate test class be created?

Reply:
Ah, got it. This new test class with the test I mentioned will make the frontend-utils dependent on at least one frontend (during testing only). But that is not an issue, I suppose.

Reply:
Yes, that should lead to a cyclic dependency. From a software engineering standpoint, the best solution would be to make this test case frontend independent! Basically, create your own mock tokens with only the interfaces provided by the frontend-utils. The references file by the tokens could be anything, e.g., pseudo-code or just Java code. This has the advantage that the unit test does not depend on additional modules, thus keeping the dependencies clean. Moreover, this makes the unit test as fine-grained as possible, ensuring that it will only fail if the tested code changes and code that is not the subject under test, for example, a language frontend.

A sufficient test input could be two files, each containing one line. For those, you could just two tokens in the test case with trivial tokens.

Reply:
Ah yes, this is surely the best way to test the functionality of the TokenPrinter. I've come up with some fun test cases*.
The only question left for me is whether I should actively check for the correct output (e.g. by comparing to a hard-coded reference output file) or just accept the output as long as no exceptions occur. The correctness could still be checked "by eye" anytime.

*I realize that the original `printFilesAsSubmissions` test from above points to an issue separate from the TokenPrinter. Is it really intentional that the `file` attribute of tokens may be empty depending on the CLI arguments? This may change also, I think, so I'd rather not add a test that _requires_ the TokenPrinter to deal with the inconsistency.

Reply:
You could just check if the printed string contains the lines of the file and the string representation of the token. This would be a relatively easy assertion to implement.

I think all tokens should point to a file, but maybe this is a weird edge case for single file submissions?

Reply:
Also, as a comment: Don't put too much time into the token printer testing, as it is obviously not as important as other parts of our codebase. A simple test or a few simple tests should be enough for now.

Reply:
That's right. Done! I'll go over the SonarCloud issues and that should be it.